### PR TITLE
Unify Dataset IO

### PR DESCRIPTION
**Pull request by TonioF**
This PR addresses #516. It ensures that, if a Configuration with Datasets is passed to a ServiceContext, each dataset is assigned to a datastore that will be used to open the data (unless the dataset is in-memory). It is attempted to find common roots for data stores to group datasets located at same/similar locations together.

When reviewing this, please consider especially the function get_data_store_pool of the ServerContext: This function used to remove a data store pool if there were no datastores explicitly configured. I had to change this, but I don't fully understand what was the original reasoning behind this.

This is a follow-up PR to erroneously merged PR #527 .

Checklist:

* [x] Add unit tests and/or doctests in docstrings
* ~[ ] Add docstrings and API docs for any new/modified user-facing classes and functions~
* ~[ ] New/modified features documented in `docs/source/*`~
* [ ] Changes documented in `CHANGES.md`
* [x] AppVeyor CI passes
* [ ] Test coverage remains or increases (target 100%)

Remember to close associated issues after merge!

### CODE REVIEW

**Review by forman — CHANGES_REQUESTED** [review]
Where are the changes in `xcube.webapi.context`?

Also I don't understand your comment: 

> When reviewing this, please consider especially the function get_data_store_pool of the ServerContext: This function used to remove a data store pool if there were no datastores explicitly configured. I had to change this, but I don't fully understand what was the original reasoning behind this.

I cannot see what you are referring to.

**Review by forman — APPROVED** [review]

File: test/webapi/test_context.py [hunk 1]
```diff
@@ -221,3 +221,246 @@ def test_interpolates_vars(self):
...
+                'Path': '/one/path/abc.zarr'
+            },
+            {
+                'Identifier': 'z_1',
+                'FileSystem': 'local',
+                'Path': '/one/path/def.zarr'
+            },
+            {
+                'Identifier': 'z_4',
+                'FileSystem': 'obs',
```
Comment (AliceBalfanz) [inline test/webapi/test_context.py:243]: I think we should stop using "obs" when referring to s3 object storage - and exchange it to "s3" as used in the DataStores configurations.

Comment (forman) [reply test/webapi/test_context.py:243]: Agreed!

Comment (TonioF) [reply test/webapi/test_context.py:243]: Correct - but for now, we'll move this to another PR.

**Pull request merged**